openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -469,45 +473,73 @@ def _verify_user(request, response, userid_in_session):
 
             if request_user_object_mismatch or session_user_mismatch:
                 # Log accumulated information stored on request for each change of user
-                if hasattr(request, 'debug_user_changes'):
-                    log.warning('An unsafe user transition was found. It either needs to be fixed or exempted.\n {}'
-                                .format('\n'.join(request.debug_user_changes)))
+                extra_logs = []
+
+                response_session_id = getattr(getattr(request, 'session', None), 'session_key', None)
+
+                # A safe-session user mismatch could be caused by the
+                # wrong session being retrieved from cache. This
+                # additional logging should reveal any such mismatch
+                # (without revealing the actual session ID in logs).
+                sessions_raw = [
+                    ('parsed_cookie', request.cookie_session_field),
+                    ('at_request', request.safe_cookie_verified_session_id),
+                    ('at_response', response_session_id),
+                ]
+                # Note that this is an ordered list of pairs, not a
+                # dict, so that the output order is consistent.
+                session_hashes = [(k, obscure_token(v)) for (k, v) in sessions_raw]
+                session_id_changed = len(set(kv[1] for kv in sessions_raw)) > 1
 
-                session_id_changed = hasattr(request.session, 'session_key') and\
-                    request.safe_cookie_verified_session_id != request.session.session_key
                 # delete old session id for security
                 del request.safe_cookie_verified_session_id
+                del request.cookie_session_field
+
+                extra_logs.append('Session changed.' if session_id_changed else 'Session did not change.')
 
-                log_suffix = 'Session changed.' if session_id_changed else 'Session did not change.'
+                # Allow comparing session IDs in both logs and metrics
+                extra_logs.append(
+                    "Hash of session ID from various sources: " +
+                    '; '.join(f'{k}={v}' for (k, v) in session_hashes)
+                )
+                for source_name, id_hash in session_hashes:
+                    set_custom_attribute(f'safe_sessions.session_id_hash.{source_name}', id_hash)
+                set_custom_attribute('safe_sessions.session_id_changed', session_id_changed)
```

Comment:
Non-blocking ideas:
1. Consider adding this data in New Relic for all transactions, which may help avoid any follow-up memcached triggered extra logging, if we get enough information across requests.
2. Consider adding annotation documentation for any new custom attributes using these [example (yet to be official) annotations](https://github.com/edx/edx-django-utils/blob/6eeca51e5a0207fc10c14a9e9bd4fff716b20f52/edx_django_utils/monitoring/internal/middleware.py#L49-L51).

Reply:
1. I think that's getting out of scope, but I'll make a note on the memcache-timer ticket to consider that.
2. Do you have a link to some in-progress guide on those new annotations? It's not obvious to me what the right annotation would be for dynamically generated attribute names.

Reply:
What I sent you **is** the in-progress guide to those annotations. :) I also wasn't sure, but I'd say just a single annotation with some notation for the dynamic part and provide details in the description. However, I could also see adding annotations for each individual setting if you thought that was needed/best.

Feel free to add this to a follow-up PR when and if you want. It is not a blocker.